### PR TITLE
codellama

### PR DESCRIPTION
**Pull request by g1y5x3**
Added and tested params for `7B` model, here is a test output
image
which matches with pytorch (128 `max_seq_len`)
image
Assume the following directory structure
```
|---LLaMA-2
|---|---7B
|---LLaMA-code
|---|---7B
|---|---7B-Python
|---|---7B-Instruct
```
If you have access to `tiny3`, weights are already downloaded under `/share/weights`. You can do the following
```
rm -r weights
ln -s /share/weights weights
```

### CODE REVIEW

**Comment by wozeparrot** [comment]
These are the args for the 34B one: `"dim": 8192, "multiple_of": 256, "n_heads": 64, "n_kv_heads": 8, "n_layers": 48, "norm_eps": 1e-05, "vocab_size": VOCAB_SIZE`
Might need to adjust vocab_size to the correct one.

Should probably also add the instruct and python variants

**Comment by g1y5x3** [comment]
Tested two more cases for `7B-Python` and `7B-Instruct`
image
image

File: examples/llama.py [hunk 1]
```diff
@@ -271,10 +307,10 @@ def quantize(tensors):
 
 class LLaMa:
   @staticmethod
-  def build(model_path, tokenizer_path, model_gen=1, model_size="7B", quantize=False):
+  def build(model_path, tokenizer_path, model_gen='1', model_size="7B", quantize=False):
     from sentencepiece import SentencePieceProcessor
     sp_model = SentencePieceProcessor(model_file=str(tokenizer_path))
-    assert sp_model.vocab_size() == VOCAB_SIZE
+    assert sp_model.vocab_size() == MODEL_PARAMS[model_gen][model_size]['args']['vocab_size']
```
Comment (prusnak) [inline examples/llama.py:313]:
```suggestion
    assert sp_model.vocab_size() == MODEL_PARAMS[model_gen][model_size]["args"]["vocab_size"]
```

File: examples/llama.py [hunk 2]
```diff
@@ -168,41 +168,77 @@ def __call__(self, tokens:Tensor, start_pos:int):
...
       "files": 4,
     },
     "65B": {
-      "args": {"dim": 8192, "multiple_of": 256, "n_heads": 64, "n_layers": 80, "norm_eps": 1e-05, "vocab_size": VOCAB_SIZE},
+      "args": {"dim": 8192, "multiple_of": 256, "n_heads": 64, "n_layers": 80, "norm_eps": 1e-05, "vocab_size": 32000},
       "files": 8,
     },
   },
-  2: {
+  '2': {
```
Comment (prusnak) [inline examples/llama.py:190]:
```suggestion
  "2": {
```

File: examples/llama.py [hunk 3]
```diff
@@ -271,10 +307,10 @@ def quantize(tensors):
 
 class LLaMa:
   @staticmethod
-  def build(model_path, tokenizer_path, model_gen=1, model_size="7B", quantize=False):
+  def build(model_path, tokenizer_path, model_gen='1', model_size="7B", quantize=False):
```
Comment (prusnak) [inline examples/llama.py:310]:
```suggestion
  def build(model_path, tokenizer_path, model_gen="1", model_size="7B", quantize=False):
```

File: examples/llama.py [hunk 4]
```diff
@@ -427,7 +517,7 @@ def greedy_until(self, prompt:str, until, max_length, temperature):
   # *** prompt engineers stop here ****
 
 
-  LLAMA_SUFFIX = {1: "", 2: "-2"}[args.gen]
+  LLAMA_SUFFIX = {'1': "", '2': "-2", "code": "-code"}[args.gen]
```
Comment (prusnak) [inline examples/llama.py:520]:
```suggestion
  LLAMA_SUFFIX = {"1": "", "2": "-2", "code": "-code"}[args.gen]
```

File: test/external/external_test_speed_llama.py [hunk 5]
```diff
@@ -28,7 +28,7 @@ def test_llama_compile(self):
     Device[Device.DEFAULT].buffer = RawFakeBuffer
 
     print("testing llama python run time")
-    model = Transformer(**MODEL_PARAMS[1]["7B"]["args"])
+    model = Transformer(**MODEL_PARAMS['1']["7B"]["args"])
```
Comment (prusnak) [inline test/external/external_test_speed_llama.py:31]:
```suggestion
    model = Transformer(**MODEL_PARAMS["1"]["7B"]["args"])
```

**Review by prusnak — CHANGES_REQUESTED** [review]
please use double quotes instead of single quotes when quoting strings for consistency with rest of the codebase

**Comment by g1y5x3** [comment]
All 7B variants are now supported and tested on `tiny3`. I was wondering if I should make a chatbot mode for codellama but all the examples they demonstrated in the [paper](https://scontent-ord5-2.xx.fbcdn.net/v/t39.2365-6/369856151_1754812304950972_1159666448927483931_n.pdf?_nc_cat=107&ccb=1-7&_nc_sid=3c67a6&_nc_ohc=wURKmnWKaloAX_9sM-1&_nc_ht=scontent-ord5-2.xx&oh=00_AfA8e08zoXYE_qgzQKoMf438Ng9xt5YE-h9jKKI-FYBj6A&oe=64F2A0CF) were basically prompt-response pairs.

**Comment by wozeparrot** [comment]
I think its fine without a chat mode. Does this support the infilling stuff?

**Comment by g1y5x3** [comment]
Not yet, infilling needs additional preprocessing/processing for the prompt. And maybe another parser argument to help trigger it. I can try adding it onto this PR.

File: examples/llama.py [hunk 6]
```diff
@@ -313,27 +350,81 @@ def greedy_until(self, prompt:str, until, max_length, temperature):
...
+output:
+Hello. I'm a 20 year old male. I'm a student at the University of Texas at Austin. I'm a sophomore majoring in Computer Science.
+
+test: 
+python3 examples/llama.py --gen='2' --temperature=0 --count=50 --prompt="Hello."
+output:
+Hello. I'm a 20 year old girl who is looking for a good lay in Palm Coast. I don't care whether it's at your place or not, as long as it's clean.
+
+test:
+python3 examples/llama.py --gen="code" --temperature=0.2 --count=50 --prompt="\
```
Comment (geohot) [inline examples/llama.py:365]: Shorter tests? I don't understand this \

File: examples/llama.py [hunk 7]
```diff
@@ -219,7 +256,7 @@ def convert(name) -> Tensor:
     disk_tensors = [model[name] for model in models]
     if len(disk_tensors) == 1 or len(disk_tensors[0].shape) == 1:
       return disk_tensors[0].to(device=Device.DEFAULT)
-    axis = 1 if name.startswith('tok_embeddings.') or name.endswith('.attention.wo.weight') or name.endswith('.feed_forward.w2.weight') else 0
+    axis = 1 if name.startswith("tok_embeddings.") or name.endswith(".attention.wo.weight") or name.endswith(".feed_forward.w2.weight") else 0
```
Comment (geohot) [inline examples/llama.py:259]: Unrelated changes

Comment (g1y5x3) [reply examples/llama.py:259]: Earlier, there was a comment about using double quotes for consistency so I just changed all of them.

File: examples/llama.py (hunk 6 above)
Comment (g1y5x3) [reply examples/llama.py:365]: I copied the "\" from the example given in their repo just to make sure we are getting the same outputs as torch. But yes, a short test example would probably be better.
